synapse/handlers/device.py

```diff
@@ -694,10 +739,18 @@ def user_device_resync(self, user_id):
         origin = get_domain_from_id(user_id)
         try:
             result = yield self.federation.query_user_devices(origin, user_id)
-        except (NotRetryingDestination, RequestSendFailed, HttpResponseException):
-            # TODO: Remember that we are now out of sync and try again
-            # later
-            logger.warning("Failed to handle device list update for %s", user_id)
+        except NotRetryingDestination:
+            # Mark the remote user's device list as stale so we know we need to retry it
+            # later.
+            yield self.store.mark_remote_user_device_cache_as_stale(user_id)
```

Comment:
I'm unsure if it matters, but it seems like the logic in here is:

1. Get a list of users who might be stale.
2. Try to sync them.
3. If successful, store the new data and mark them as not stale (fresh?).
4. If failure, mark them as stale.

In other cases step 1 above is skipped (during incoming device updates).

I'm unsure if this is worth worrying about, but it seems kind of wasteful from a database POV.

Reply:
Agreed, again i'm unsure the current logic here is the right one. Though it sounds like refactoring part of `user_device_resync` could solve most issues, I'll look into that.

Reply:
You might be able to pass a `mark_as_stale_on_failure` (better name needed ™️ ) flag into `user_device_resync` which is usually `True` except when called from `_maybe_retry_device_resync`?

Would be a bit annoying, but pretty straightforward IMO.

Reply:
Oh, that's actually a pretty good solution that's both simple and doesn't need changing the other uses of that function, I think I'll do that, thanks!

Reply:
Done in https://github.com/matrix-org/synapse/pull/7453/commits/70d17c53dcd1a0dd69a3fc677d0f191db1bf154c